### PR TITLE
Implement Chaosnet broadcast sending

### PR DESCRIPTION
**Pull request by bictorv**
ITS learns how to send BRD packets. Note that it is slightly different from the spec (MIT AIM 628 Sec 4.5) in that BRD packets are retransmitted by ITS, and only ANS responses are delivered to the user, until the user closes the channel. This requires some discipline of the user, to use reasonably short timeouts and closing the channel. Often you also need to use an interrupt routine to detect incoming packets, since NETBLK isn't useful.

Example programs are included: CHATST gets a new "BRD" command, and NAME learns to use broadcast for the `/L` switch ("associated lisp machines") and for the `*LISPM` ("all lispms") JCL (note: not `@*LISPM`). The `/L` switch uses broadcast on the local subnet, while `*LISPM` uses global broadcast. Additionally, NAMDRG uses broadcast to find lispm users and free lisp machines (see #2195 ).

PEEK learns to show the Broadcast-sent (or BRDSNT) state of Chaosnet connections.

(For klh10 using ifmeth=chudp, the most recent version of klh10 is required in order for broadcast packets to actually be sent on chudp links. For ifmeth=pcap, it works also on older versions.)

Future work: 
- implement broadcast `TIME` checking
- teach NAME to show free lisp machines
- teach NAME to broadcast using the `LOAD` protocol, and for responses with `Users: n` for n>0, connect with the `NAME` protocol and show who's on.

Acknowledgement: Many thanks to @larsbrinkhoff for the new `mlftp` program, which increased productivity! (See branch of https://github.com/Chaosnet/chaosnet-tools.)

### CODE REVIEW

File: src/sysen2/peek.631 [hunk 1]
```diff
@@ -3656,16 +3656,17 @@ KAOS00:	ONUM 3,C
...
 	6TYPE (A)[SIXBIT/CLOSED/
 		  SIXBIT/LISTEN/
 		  SIXBIT/RFCRCV/
 		  SIXBIT/RFCSNT/
 		  SIXBIT/OPEN  /
 		  SIXBIT/LOS   /
 		  SIXBIT/INCXMT/
-		  SIXBIT/LOWLVL/
+		  SIXBIT/FORIGN/ ;Foreign protocol mode
+		  SIXBIT/BRDSNT/ ;Broadcast sent
```
Comment (larsbrinkhoff) [inline src/sysen2/peek.631:3669]: You added two more strings, but only increased the number by one.  Is that right?

Comment (bictorv) [reply src/sysen2/peek.631:3669]: Previously the state %CSFRN (foreign protocol mode) would be presented as "LOWLVL" which seems arachaic (FORIGN is clearer, in my mind), while anything over %CSFRN was presented as "GARBAG". With the new code, %CSBRS (which is %CSFRN+1) is "BRDSNT" and anything higher is "GARBAG".

Comment (larsbrinkhoff) [reply src/sysen2/peek.631:3669]: Oh, right!  Never mind then.

**Comment by bictorv** [comment]
> Very nice!
> 
> > copy code from NAME to NAMDRG, to solve issue #2195
> 
> My only known issue with NAMDRG not finding free Lispms is that there are none on the network. Can you explain how NAMDRG worked or didn't work before, and how broadcasting would impact that?

Without broadcast, NAMDRG would need to know (hardwired in the code) all lispms that might be free or not. With broadcast, they respond whereever they are (with up-to-date code running, which is now "years old").

**Comment by bictorv** [comment]
But I'm seriously considering marking this PR as "draft" until I played^Wexperiemented more with manual retransmissions. It might not be as awkward as I imagined.

**Comment by bictorv** [comment]
> But I'm seriously considering marking this PR as "draft" until I played^Wexperiemented more with manual retransmissions. It might not be as awkward as I imagined.

There. It wasn't all that awkward.

**Comment by bictorv** [comment]
Does the simh implementation of CH11 (and CH10?) implement broadcast over CHUDP?

**Comment by larsbrinkhoff** [comment]
No, the SIMH variants don't do any broadcasting.

**Comment by bictorv** [comment]
> No, the SIMH variants don't do any broadcasting.

But if I read the docs correctly, all it does is broadcasting - i.e. it sends all packets to its (single) peer? The code doesn't seem to even look at the destination address. So then this PR "should work" also on simh-based systems? Anyone, please try?

**Comment by larsbrinkhoff** [comment]
In that case, yes of course they all do broadcast!

@eswenson1 you have several SIMH emulators on a single Chaosnet segment, right?

**Comment by eswenson1** [comment]
Yes, I have KLH10 (2) and one each of pdp10-kl, ka, and ks.

what ca I do to test? (Or feel free to use them).

**Comment by bictorv** [comment]
> Yes, I have KLH10 (2) and one each of pdp10-kl, ka, and ks.
> 
> what ca I do to test? (Or feel free to use them).

If you use the new `system;chaos >` to build a new ITS, boot it, then compile the new `chatst`, and use **b**RD **FINGER** and see if you get replies?

**Comment by eswenson1** [comment]
On EXA (pdp10-ka), I get:

```
EXA>:chatst

>BRD Contact name: finger
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"
 Sending...
 Have packets: 0
 Timed out
 Closing and reopening channels

>
```

**Comment by eswenson1** [comment]
On EXL (pdp10-kl), I get:
```
EXL>:chatst

>BRD Contact name: finger
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"

*** IOC ERROR *** Channel=13, PC=4037
State is now Closed
Input count=0., Output count=0.
Input window=8., Output window=0.
Dismissing...
ERROR: PKTIOT:  DEVICE FULL
4037>>.CALL 6062 (PKTIOT)
```

**Comment by eswenson1** [comment]
On EXS (pdp10-ks), I get:
```
:chatst

>BRD Contact name: finger
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"
 Sending...
 Have packets: 0
 Timed out
 Closing and reopening channels

>
```

**Comment by eswenson1** [comment]
For EX (klh10), I get:
```
chatst↑K!

>BRD Contact name: finger
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 38.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························finger"
 Sending...
 Have packets: 0
 Timed out
 Closing and reopening channels

>
```
Note: I have not updated KLH10 with the patch you referenced somewhere.  Do I need/want to?

**Comment by eswenson1** [comment]
I should have finger servers on all those machines. However, since your instructions said to rebuild ITS, build the updated CHARST, and run it, I hadn’t rebuilt NAME/PEEK and neither had been purified and updated against the new ITS. 

I’ve since updated NAME and PEEK. Should I rerun the test?

**Comment by bictorv** [comment]
It's confusing, but the Chaosnet FINGER protocol is not what is used by :FINGER (:F, :NAME etc) - it uses the NAME protocol. FINGER is used by Lisp machines (and my iMac).

Please rerun the test, but with the LOAD contact name instead?

**Comment by eswenson1** [comment]
I reran on EXA.  Got this:
IMG_3012

**Review by larsbrinkhoff — CHANGES_REQUESTED** [review]
Before merging the final version (maybe the current version!), let's rearrange things a bit.  Maybe squash all into a single commit, or maybe make it one for ITS and another for the tools?

**Comment by bictorv** [comment]
> I reran on EXA. Got this:

Mysterious. After trying to build a pdp10-ka ITS by cloning the repo and doing `make EMULATOR=pdp10-ka all`, which stops with `The last command timed out.`, I realised `make EMULATOR=pdp10-ka download` was a lot easier. Then I configured a Chaosnet address on my subnet 7, copied `system;chaos >` from BV, rebuilt ITS and rebooted, and copied `bv;chatst >` from BV, compiled it and ran it. 

It seems to work fine. I think you have bad luck? ;-) But maybe I should try configuring a new subnet and see what happens.

At some point I will give some feedback on the build process, and patch ITS to handle the situation where the CH10 "CHAOSNET ADDRESS SWITCHES ARE SET WRONG" (and, like for CH11, simply use the address from the interface, so you don't have to recompile ITS to change addresses).

**Comment by eswenson1** [comment]
I tried on EXS and got this:
```
chatst↑K!

>BRD Contact name: load
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 Timed out
 Closing and reopening channels

>
```

**Comment by eswenson1** [comment]
And I tried on EXL and got this:
```
↑Kchatst!

>BRD Contact name: load
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 Timed out
 Closing and reopening channels

>
```

**Comment by eswenson1** [comment]
Retried on EXA and got this:
```
chatst↑K!

>BRD Contact name: load
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 More packets: 0
 Now sending: OP 16 BRD, FC 0., NB 36.
From 0-0 to 0-0
Pkt # 0, ack # 40
Data: "······························load"
 Sending...
 Have packets: 0
 Timed out
 Closing and reopening channels

>
```
So I get the same results on all of pdp10-ka, pdp10-kl, and pdp10-ks.  What is the output supposed to look like?

**Comment by bictorv** [comment]
I just logged in on EXA and compiled `SYSEN3;CHATST` to `.temp.`, and when I run it saying BRD **LOAD** (in uppercase!) I get responses. Try it (it also works with `:CHATST`, running `SYS2;TS CHATST`).

It turns out ITS assumes contact names in incoming RFC and BRD packets are in uppercase, but it doesn't _make_ them uppercase if they aren't. So you need to write it in uppercase. Of course CHATST should make them uppercase.

**Comment by bictorv** [comment]
> Before merging the final version (maybe the current version!), let's rearrange things a bit. Maybe squash all into a single commit, or maybe make it one for ITS and another for the tools?

You know how you want it and how to make it be so, so go ahead? Now that I have a PDP10-KA, I might make NAMDRG use broadcast, but probably not until next week.

**Comment by eswenson1** [comment]
> It turns out ITS assumes contact names in incoming RFC and BRD packets are in uppercase, but it doesn't _make_ them uppercase if they aren't. So you need to write it in uppercase. Of course CHATST should make them uppercase.

Well, *that* makes a difference.  Now I responses on EXA, EXS, and EXL.  I haven’t updated KLH10 yet for ES and EX, so I don’t expect good responses there.

**Comment by eswenson1** [comment]
@bictorv I just rebuilt klh10 from the master branch (4955fb89b6eea60da5ba6fdfae737f5bc1c227d1) of the git@github.com:PDP-10/klh10.git repo.  I note that there were differences, credited to you, over what I had been running.  However, the last commit (by you) was:
```
commit 4955fb89b6eea60da5ba6fdfae737f5bc1c227d1
Author: Björn Victor <bjorn@victor.se>
Date:   Mon Jan 16 12:16:12 2023 +0100

    Amend and correct 602e733
```
Note the date is Jan 16, 2023.  So this must not be the correct place for the changes.  Where did you make the KLH10 changes?

**Comment by eswenson1** [comment]
I updated KLH10 with your latest changes, and now EX, also, returns good data on a CHATST with "b LOAD".  I'll update ES next.

**Comment by bictorv** [comment]
Now NAMDRG uses broadcast, too. (And shows the whole location of free lispm instead of stopping at the first space.)

**Review by larsbrinkhoff — APPROVED** [review]
Bumped NAMDRG version.

**Review by eswenson1 — COMMENTED** [review]
I noticed, @larsbrinkhoff, that you didn’t bump the version of SYSTEM;CHAOS. Is that because we did in the other PR? Ideally FN2 should increment after every PR, in my mind.

**Comment by larsbrinkhoff** [comment]
Right, these were merged in quick succession so I figured it would not cause a problem to anyone if the version wasn't bumped again.

**Pull request merged**